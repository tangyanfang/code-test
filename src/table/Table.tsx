/**
 * Created by uedc on 2021/10/11.
 */

import { defineComponent, h } from '@vue/composition-api'
import TableHeader from './table_header';
import TableBody from './table_body';
import TableFooter from './table_footer';

export default defineComponent({
  name: 'Table',
  components: {
    TableHeader,
    TableBody,
    TableFooter,
  },
  props: {
    tableData: {
      type: Array,
      default: () => {
        return []
      }
    }
  },
  setup(props, { slots }) {
    window.console.log(props.tableData);
    return () => {
      return (
        <table border="1" class='my-table'>
          <table-header></table-header>
          <table-body></table-body>
          {/* <table-footer></table-footer> */}
        </table>
      )
    }
  },
  render() {
    return h(
      'table',
      {
        class: 'my-table',
        cellspacing: '0',
        cellpadding: '0',
        border: '1',
      },
      [
        h('table-header', {
          props: {
            tableTitleList: []
          }
        }),
        h('<table-body', {

        })
      ]
    )
  }
})
