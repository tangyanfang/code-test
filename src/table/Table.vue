<template>
  <table class="my-table">
    <table-header ref="tableHeaderComponent"
                  :header-data="header"
                  @sort-handle="sortHandleEmit"
    />
    <table-body ref="tableBodyComponent"
                :header-data="header"
                :content-data="content.list"
                :page-size="size"
    />
    <table-pager
      :max-page="maxPage"
      :page-size="size"
      @page-handle="pageHandleEmit"
      @init-sort-type="initSortTypeEmit"
    />
  </table>
</template>

<script lang="ts">

/**
 * Created by uedc on 2021/10/11.
 */

import { defineComponent, reactive, ref, computed } from '@vue/composition-api'
import TableHeader from './table_header/index.vue'
import TableBody from './table_body/index.vue'
import TablePager from './pager/index.vue'
import { PropType } from '@vue/runtime-dom';
import { row, columnTitle } from './types';

export default defineComponent({
  name: 'Table',
  components: {
    TableHeader,
    TableBody,
    TablePager,
  },
  props: {

    // 表头内容
    headerData: {
      type: Array as PropType<Array<columnTitle>>,
      default: () => {
        return []
      },
    },

    // 表格内容
    contentData: {
      type: Array as PropType<Array<row>>,
      default: () => {
        return []
      },
    },

    // 每页展示的数据条数
    pageSize: {
      type: Number,
      default: 20,
    },
  },
  setup(props) {
    const header = props.headerData
    const content = reactive({
      list: props.contentData,
    })
    const size = ref(props.pageSize)
    const tableBodyComponent = ref<any>()
    const tableHeaderComponent = ref<any>()

    // 向body组件发送排序事件
    const sortHandleEmit = (columnId: string, type: string) => {
      tableBodyComponent.value.sortHandle(columnId, type)
    }

    // 向body组件发送分页事件
    const pageHandleEmit = (currentPage: number, currentSize: number) => {
      size.value = currentSize;
      tableBodyComponent.value.pageHandle(currentPage, currentSize)
    }

    // 向head组件发送初始化排序方式事件
    const initSortTypeEmit = () => {
      tableHeaderComponent.value.initSortType()
    }

    // 获取最大页码
    const maxPage = computed(() => Math.ceil(content.list.length/size.value));


    return {
      header,
      content,
      size,
      maxPage,
      sortHandleEmit,
      pageHandleEmit,
      initSortTypeEmit,
      tableBodyComponent,
      tableHeaderComponent,
    }
  },
})
</script>
