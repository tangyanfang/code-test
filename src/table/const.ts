// 常量定义

// 排序方式常量
export const SORT_TYPES = {
    asc: 'asc',
    des: 'des',
    reset: 'reset'
};
export const ASC_CLS = 'ascending';
export const DES_CLS = 'descending';