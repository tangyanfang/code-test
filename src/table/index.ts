/**
 * Created by uedc on 2021/10/11.
 */

import './style/index.less'
import TestTable from './Table.vue'

export { TestTable }
