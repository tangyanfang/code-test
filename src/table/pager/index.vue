<template>
  <div class="my-table-pager">
    <span>选择每页展示的数据条数：</span>
    <select v-model="currentSize"
            class="size-select">
      <option v-for="item in pageSizes"
              :key="item"
              :value ="item">{{item}}</option>
    </select>
    <ul class="pager-list">
      <li v-for="pager in pagers"
          :key="pager"
          :class="{ active: currentPage === pager }"
          class="number"
          @click="onPagerClick(pager)"
      >
        {{ pager }}
      </li>
    </ul>
  </div>
  
</template>
<script lang="ts">

/**
 * 分页组件
 */

import { defineComponent, reactive, toRefs, watch, computed } from '@vue/composition-api'
import debug from 'debug';
const log = debug('table:pager');

export default defineComponent({
  name: 'TablePager',
  props: {
    total: {
      type: Number,
      default: 0,
    },
  },
  emits: ['page-handle'],
  setup(props, {emit}) {

    // 基本数据
    const option = reactive({
      pageSizes: [20, 50, 100],
      currentSize: 20,
      currentPage: 1
    });
    const useConvertProps = (props: any) => {
      const maxPage =  computed(() => {
        return Math.ceil(props.total/option.currentSize);
      });

      // 获取所有页码
      const pagers = computed(() => Array.from(Array(maxPage.value)).map((v, i) => i+1)); 
      return  { maxPage, pagers };
    };
    const { maxPage, pagers } = useConvertProps(props);

    // 监听maxPage变化
    watch(maxPage, (newV) => {
      if (newV < option.currentPage) {
        option.currentPage = 1;
      }
    });

    // 监听currentSize变化
    watch(() => option.currentSize, (newV) => {
      changePage(option.currentPage, newV);
    });

    // 监听currentPage变化
    watch(() => option.currentPage, (newV) => {
      changePage(newV, option.currentSize);
    });

    // 选择页码
    const onPagerClick = (page: number): void => {
      option.currentPage = page;
    };

    // 抛出分页事件
    const changePage = (val: number, currentSize: number) => {
      emit('page-handle', val, currentSize)
    };
    return { 
      ...toRefs(option), 
      pagers,
      maxPage, 
      useConvertProps,
      onPagerClick, 
      changePage
    }
  },
})
</script>
