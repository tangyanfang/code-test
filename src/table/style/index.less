// 表格公共样式

.my-table {
    margin: auto;
    border: 1px solid red;
}
.my-table-header {
    border-bottom: 1px solid red;
    tr {
        button {
            background-color: red;
            padding: 0 4px;
        }
    }
}
.my-table-body {
    height: 600px;
    overflow-y: auto;
}
th {
    border: 1px solid red;
    padding: 4px;
}
.row-content {
    td {
        display: inline-block;
        padding: 4px 16px;
    }
}
.my-table-pager {
    display: flex;
    justify-content: flex-end;
    border-top: 1px solid red;
    padding-top: 2px;
    .pager-list {
        display: flex;
    }
    li {
        width: 20px;
        text-align: center;
        border: 1px solid #ccc;
        cursor: pointer;
    }
    .active {
        border: 1px solid red;
    }
    .size-select {
        width: 50px;
        background-color: aqua;
        opacity: 0.5;
        margin-right: 10px;
    }
}