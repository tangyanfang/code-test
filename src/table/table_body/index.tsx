/**
 * è¡¨æ ¼body
 */

 import { defineComponent } from '@vue/composition-api'
 import { tableProps } from '../types'
 
 export default defineComponent({
   name: 'TableBody',
   props: tableProps,
   setup(props, { slots }) {
     return () => {
       return (
         <div class='my-table__body'>
           <tr>
            <td>1</td>
            <td>1</td>
          </tr>
          <tr>
            <td>2</td>
            <td>2</td>
          </tr>
         </div>
       )
     }
   },
 })
 
