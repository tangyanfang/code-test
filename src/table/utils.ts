// 工具方法

// 排序
export let compare = (property: string, type: string) => {
  return function(a: Object ,b: Object){
    const value1 = a[property]
    const value2 = b[property]
    if (type === 'asc') {
      return value1 - value2;
    }
    return value2 - value1;
  }
}

// 表格数据处理
export let getColumns = (row: any, header: any) => {
  const handleColumns: any = []
  header.forEach((item: any) => {
    handleColumns.push({
      id: item.columnProp,
      value: row[item.columnProp],
      scale: item.scale || 1,
    })
  })
  return handleColumns;
}
export let hanledRowData = <T>(tableData: Array<T>, columnTitle: any): Array<T> => {
  return tableData.map((row: any) => {
    return {
      ...row,
      columns: getColumns(row, columnTitle),
    }
  })
}

// 获取每个td的宽度方法
export let getTdWidth = (scale: number, column: any) => {
  return `${scale/(column.length) * 100}%`
}
