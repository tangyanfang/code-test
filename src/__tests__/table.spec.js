import { mount } from '@vue/test-utils'
import Vue from 'vue'
import CompositionApi from '@vue/composition-api'
import { TestTable } from '../table'
import TableHeader from '../table/table_header/index.vue'
import TableBody  from '../table/table_body/index.vue'
import TablePager from '../table/pager/index.vue'
import { headerData, column1DesSortData, column1AscSortData,  sortData} from '../../demo/test_data'
import { compare } from '../table/utils';

Vue.use(CompositionApi)

describe('Table', () => {
  const TableMount = options => mount(TestTable, options)
  const tableHeader = options => mount(TableHeader, options)
  const tableBody = options => mount(TableBody, options)
  const tablePager = options => mount(TablePager, options)

  test('render', () => {
    const wrapper = TableMount()
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => {
      wrapper.vm.$forceUpdate()
      wrapper.vm.$destroy()
    }).not.toThrow()
  })

  // 排序测试
  test('排序测试', async () => {
    const tableHeaderWrapper = tableHeader({
      propsData: {
        headerData: headerData
      }
    });
    // 期待存在排序按钮
    expect(tableHeaderWrapper.find('button').exists()).toBe(true)

    // 模拟点击第1列降序
    tableHeaderWrapper.findAll('button').wrappers[1].trigger('click') 

    // 点击完降序按钮后，期待按钮置灰
    expect(tableHeaderWrapper.findAll('button').wrappers[1].exists('[disabled=disabled]')).toBeTruthy()

    //断言是否发出排序事件
    expect(tableHeaderWrapper.emitted()['sort-handle']).toBeTruthy()
  })

  // 分页测试
  test('分页测试', async () => {
    const tablePagerWrapper = tablePager({
      propsData: {
        maxPage: 5
      }
    });

    // 测试是否默认激活第一页
    expect(tablePagerWrapper.findAll('li.number').wrappers[0].find('.active').exists()).toBeTruthy()

    // 模拟点击第3页
    await tablePagerWrapper.findAll('li.number').wrappers[2].trigger('click') 

    // 测试分页条中的3是否激活
    expect(tablePagerWrapper.findAll('li.number').wrappers[2].find('.active').exists()).toBeTruthy()

    // 测试当前页码大于最大页码时，是否跳转到了第一页
    await tablePagerWrapper.setData({ currentPage: 5 })
    await tablePagerWrapper.setProps({ maxPage: 3 })
    expect(tablePagerWrapper.findAll('li.number').wrappers[0].find('.active').exists()).toBeTruthy()

    // 测试currentSize变化时，有发送分页事件
    await tablePagerWrapper.setData({ currentSize: 50 })
    expect(tablePagerWrapper.emitted()['page-handle']).toBeTruthy()
  })

  //
  test('table-body排序测试', async () => {
    const table = TableMount({
      propsData: {
        headerData: headerData,
        contentData: sortData,
        pageSize: 20
      }
    })

    // tableHeader发送排序事件给table，table调用table-body的排序方法
    // 升序
    await table.findAll('.my-table-header th').wrappers[0].trigger('click')
    expect(table.findComponent(TableHeader).emitted()['sort-handle']).toBeTruthy()
    let ascData = table.findComponent(TableBody).findAll('.my-table-body .column1').wrappers.map(node => Number(node.element.innerHTML))
    let ascDataBeforeStr = ascData.toString()
    expect(ascDataBeforeStr).toEqual(column1AscSortData.sort((a, b) => a - b).toString())

    // 降序
    await table.findAll('.my-table-header th').wrappers[0].trigger('click')
    let desData = table.findComponent(TableBody).findAll('.my-table-body .column1').wrappers.map(node => Number(node.element.innerHTML))
    let desDataBeforeStr = desData.toString()
    expect(desDataBeforeStr).toEqual(column1DesSortData.sort((a, b) => b - a).toString())
  })

  // 测试排序方法是否正确
  test('compare方法测试', async () => {
    let desSortData = sortData.sort(compare('column1', 'des'));
    expect(desSortData.toString()).toEqual(desSortData.toString())
    let ascSortData = desSortData.sort(compare('column1', 'asc'));
    expect(ascSortData.toString()).toEqual(sortData.toString())
  })
})
