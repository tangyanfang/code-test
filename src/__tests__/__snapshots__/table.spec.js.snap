// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table render 1`] = `
"<table class=\\"my-table\\">
  <div class=\\"my-table-header\\">
    <tr></tr>
  </div>
  <div class=\\"my-table-body\\">
    <div>暂无数据</div>
  </div>
  <div class=\\"my-table-pager\\" page-size=\\"20\\"><span>选择每页展示的数据条数：</span> <select class=\\"size-select\\">
      <option value=\\"20\\">20</option>
      <option value=\\"50\\">50</option>
      <option value=\\"100\\">100</option>
    </select>
    <ul class=\\"pager-list\\"></ul>
  </div>
</table>"
`;
